src/ints/mouse_dos_driver.cpp

```diff
@@ -0,0 +1,1440 @@
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "mouse.h"
+#include "mouse_core.h"
+
+#include <algorithm>
+
+#include "bios.h"
+#include "bitops.h"
+#include "callback.h"
+#include "checks.h"
+#include "cpu.h"
+#include "dos_inc.h"
+#include "int10.h"
+#include "pic.h"
+#include "regs.h"
+
+using namespace bit::literals;
+
+CHECK_NARROWING();
+
+// This file implements the DOS mouse driver interface,
+// using host system events
+
+// Reference:
+// - Ralf Brown's Interrupt List
+// - WHEELAPI.TXT from CuteMouse package
+// - https://www.stanislavs.org/helppc/int_33.html
+// - http://www2.ift.ulaval.ca/~marchand/ift17583/dosints.pdf
+
+#define CURSOR_SIZE_X  16
+#define CURSOR_SIZE_Y  16
+#define CURSOR_SIZE_XY (CURSOR_SIZE_X * CURSOR_SIZE_Y)
+
+#define NUM_BUTTONS 3
+
+enum class MouseCursor : uint8_t { Software = 0, Hardware = 1, Text = 2 };
+
+// These values represent 'hardware' state, not driver state
+
+MouseButtons12S buttons = 0;
+float pos_x     = 0.0f;
+float pos_y     = 0.0f;
+int16_t wheel   = 0;
+uint8_t rate_hz = 0; // TODO: add proper reaction for 0 (disable driver)
+
+static struct { // DOS driver state
+
+    // Structure containing (only!) data which should be
+    // saved/restored during task switching
+
+    // DANGER, WILL ROBINSON!
+    ///
+    // This whole structure can be read or written from the guest side
+    // via virtual DOS driver, functions 0x15 / 0x16 / 0x17.
+    // Do not put here any array indices, pointers, or anything that
+    // can crash the emulator if filled-in incorrectly, or that can
+    // be used by malicious code to escape from emulation!
+
+    bool enabled    = false; // TODO: make use of this
+    bool cute_mouse = false;
+
+    uint16_t times_pressed[NUM_BUTTONS]   = {0};
+    uint16_t times_released[NUM_BUTTONS]  = {0};
+    uint16_t last_released_x[NUM_BUTTONS] = {0};
+    uint16_t last_released_y[NUM_BUTTONS] = {0};
+    uint16_t last_pressed_x[NUM_BUTTONS]  = {0};
+    uint16_t last_pressed_y[NUM_BUTTONS]  = {0};
+    uint16_t last_wheel_moved_x           = 0;
+    uint16_t last_wheel_moved_y           = 0;
+
+    float mickey_x = 0.0f;
+    float mickey_y = 0.0f;
+
+    float mickeys_per_px_x = 0.0f;
+    float mickeys_per_px_y = 0.0f;
+    float pxs_per_mickey_x = 0.0f;
+    float pxs_per_mickey_y = 0.0f;
+
+    uint16_t gran_x = 0; // granularity mask
+    uint16_t gran_y = 0;
+
+    int16_t update_region_x[2] = {0};
+    int16_t update_region_y[2] = {0};
+
+    uint16_t language = 0; // language for driver messages, unused
+    uint8_t mode      = 0;
```

Comment:
Could mode be turned into a typed enum?

Reply:
Probably not - if I'm not mistaken, it is being read from guest system memory, using `mem_readb(BIOS_VIDEO_MODE);`. I'm not even sure what it is used for.